joomla/joomla-cms
# [4.0] Fix errors in installation when specifying a not existing database

Pull Request for Issue #27924 .

### Summary of Changes

#### What this PR changes and fixes:

- Let function CreateDatabase of the database model enqueue messages and return false in case of errors, like it is expected by the calling functions, and not throw exceptions. This fixes issue #27924 .
- Move variables for CMS requirements for minimum database server versions which have been introduced with PR #28135 from the setup model to the database helper and added a function to get the relevant value for the particular database type, which is the maximum of either the database driver's or the CMS' requirement.
- Let both setup and database model use the same check for the minimum required version from the database helper. The setup model checks this if the database already exists, and the database model will check that later when trying to create a database.
- Added the check for database encryption capabilities of the server to the database model for the same case (newly created database).

#### What this PR doesn't fix:

Depending on the particular database driver, exception texts in error messages are more or less meaningful. This PR doesn't change that.

### Testing Instructions

_Step 1:_ Start a new installation of a current 4.0-dev **without** the patch of this PR applied.

_Step 2:_ When coming to the database information, enter a valid server and a valid user name and password for a user who doesn't have the privilege to create databases. Enter a database name for which no database exists. When having filled out all fields, use the "Install Joomla >" button to start the installation.

_Result:_ Depending on the particular database driver you see a more or less meaningful error message.

_Step 3:_ Correct the database name so it fits to an existing database for which the previously entered user has all privileges required for a Joomla installation. Then use the "Install Joomla >" button again.
Result: See issue #27924 and section "Actual result" below.

_Step 4:_ Delete configuration.php, and either delete the session cookies in your browser settings or close wour browser window and open it again.

_Step 5:_ Apply the patch of this PR.

_Step 6:_ Start a new installation of a current 4.0-dev **with** the patch of this PR applied.

_Step 7:_ Repeat steps 2 and 3 above.

_Result:_ Installation continues with the corrected database name, see section "Expected result" below.

_Step 8:_ Delete again configuration.php, and again either delete the session cookies in your browser settings or close wour browser window and open it again.

_Step 9:_ Start again a new installation of a current 4.0-dev **with** the patch of this PR applied.

_Step 10:_ When coming to the database information, test that the version checks introduced with PR #28135 still work, i.e. either use a databas server which doesn't fulfill the minimum version requirement by either the database driver or the CMS, or patch either the database driver's or the CMS' requirement so it isn't fulfilled by your database server.

The CMS's requirements for the particular database types can be found and if necessary patched here:
https://github.com/joomla/joomla-cms/blob/c90cbb57b15e464df0d402e1db990aa843d1569c/installation/src/Helper/DatabaseHelper.php#L23-L48

_Result:_ The check for the minimum required database server version still works for any kind of database. If the requirement is not fulfilled, it is possible to change to another database which fulfills the requirement (or patch the requirement) and continue with the installation.

_The next test steps 11 to 13 currently only work (with or without this PR) for MySQLi and MySQL (PDO) but not for PostgreSQL (PDO). I have no idea how it is with MariaDB. PostgreSQL users please skip this test and count is as successful. I will fix installation of new database for PostgreSQL with a later PR._

_Step 11:_ Delete again configuration.php, and again either delete the session cookies in your browser settings or close wour browser window and open it again.

_Step 12:_ Start again a new installation of a current 4.0-dev **with** the patch of this PR applied.

_Step 13:_ When coming to the database information, enter a valid server and a valid user name and password for a user who has the privilege to create databases, e.g. "root" on MySQL. Enter a database name for which no database exists. When having filled out all fields, use the "Install Joomla >" button to start the installation.

_Result:_ Creating the database works as well as before for MySQLi and MySQL (PDO).

### Expected result

When entering a not existing database as database name in the installation form and a user who doesn't have the permission to create a database, you can correct database or user information after the error message and then continue with the installation.

### Actual result

When entering a not existing database as database name in the installation form and a user who doesn't have the permission to create a database, and then after the error message correct database or user information and start the installation again, it stops with an unspecified database error, see issue #27924 .

### Documentation Changes Required

None.
